### ISSUE TITLE
Speeds up compiling cases-insensitive regular expressions

### ISSUE THREAD

**Issue by serhiy-storchaka** [python/cpython#74463]
BPO | [30277](https://bugs.python.org/issue30277)
--- | :---
Nosy | @ezio-melotti, @serhiy-storchaka
PRs | <li>python/cpython#1468</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = 'https://github.com/serhiy-storchaka'
closed_at = <Date 2017-05-05.07:43:31.387>
created_at = <Date 2017-05-05.07:03:04.973>
labels = ['expert-regex', '3.7', 'library', 'performance']
title = 'Speeds up compiling cases-insensitive regular expressions'
updated_at = <Date 2017-05-05.07:43:31.387>
user = 'https://github.com/serhiy-storchaka'
```

bugs.python.org fields:
```python
activity = <Date 2017-05-05.07:43:31.387>
actor = 'serhiy.storchaka'
assignee = 'serhiy.storchaka'
closed = True
closed_date = <Date 2017-05-05.07:43:31.387>
closer = 'serhiy.storchaka'
components = ['Library (Lib)', 'Regular Expressions']
creation = <Date 2017-05-05.07:03:04.973>
creator = 'serhiy.storchaka'
dependencies = []
files = []
hgrepos = []
issue_num = 30277
keywords = []
message_count = 2.0
messages = ['293049', '293062']
nosy_count = 3.0
nosy_names = ['ezio.melotti', 'mrabarnett', 'serhiy.storchaka']
pr_nums = ['1468']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'performance'
url = 'https://bugs.python.org/issue30277'
versions = ['Python 3.7']
```

</p></details>

**Comment by serhiy-storchaka** [comment]
Currently \_sre.getlower() takes two arguments. Depending on the bits set in the second argument it uses one of three algorithms for determining the lower case of the character -- Unicode, ASCII-only, and locale-depended. After resolving bpo-30215 \_sre.getlower() no longer used for locale-depended case. Proposed patch replaces \_sre.getlower() with two one-argument functions: \_sre.ascii_tolower() and \_sre.unicode_tolower(). This slightly speeds up compiling cases-insensitive regular expressions, especially containing ranges.

```sh
$ ./python -m timeit -s 'import sre_compile'  'sre_compile.compile("(?i)ABCDEFGHIJKLMNOPQRSTUVWXYZ", 0)'
Unpatched:  2000 loops, best of 5: 180 usec per loop
Patched:    2000 loops, best of 5: 173 usec per loop

$ ./python -m timeit -s 'import sre_compile'  'sre_compile.compile("(?ia)ABCDEFGHIJKLMNOPQRSTUVWXYZ", 0)'
Unpatched:  2000 loops, best of 5: 175 usec per loop
Patched:    2000 loops, best of 5: 168 usec per loop

$ ./python -m timeit -s 'import sre_compile'  'sre_compile.compile("(?i)[A-Z]", 0)'
Unpatched:  500 loops, best of 5: 788 usec per loop
Patched:    500 loops, best of 5: 766 usec per loop

$ ./python -m timeit -s 'import sre_compile'  'sre_compile.compile("(?ia)[A-Z]", 0)'
Unpatched:  5000 loops, best of 5: 92 usec per loop
Patched:    5000 loops, best of 5: 83.2 usec per loop

$ ./python -m timeit -s 'import sre_compile'  'sre_compile.compile("(?i)[\u0410-\u042f]", 0)'
Unpatched:  2000 loops, best of 5: 141 usec per loop
Patched:    2000 loops, best of 5: 122 usec per loop

$ ./python -m timeit -s 'import sre_compile'  'sre_compile.compile("(?i)[\u0000-\uffff]", 0)'
Unpatched:  5 loops, best of 5: 59 msec per loop
Patched:    10 loops, best of 5: 28.9 msec per loop
```

**Comment by serhiy-storchaka** [comment]
New changeset 7186cc29be352bed6f1110873283d073fd0643e4 by Serhiy Storchaka in branch 'master':
bpo-30277: Replace \_sre.getlower() with \_sre.ascii_tolower() and \_sre.unicode_tolower(). (bpo-1468)
https://github.com/python/cpython/commit/7186cc29be352bed6f1110873283d073fd0643e4
